include/ensmallen_bits/moead/moead_impl.hpp

```diff
@@ -0,0 +1,358 @@
+{
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  if(lowerBound.size() != upperBound.size())
+  {
+    throw std::logic_error("MOEAD::Optimize(): size of lowerBound and upperBound "
+        "must be equal.");
+  }
+  if(lowerBound.size() != iterate.n_elem)
+  {
+    throw std::logic_error("MOEAD::Optimize(): there should be a lower bound and "
+        "an upper bound for each variable in the initial point.");
+  }
+  // Number of objective functions. Represented by m in the paper.
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // 1.1 The external population, non-dominated solutions.
+  std::vector<MatType> externalPopulation;
+  std::vector<arma::vec> externalPopulationFValue;
+
+  // Weight vectors, where each one of them represents a decomposition.
+  std::vector<arma::vec> weights(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    weights[i] = arma::vec(numObjectives, arma::fill::randu);
+
+  // 1.2 Storing the indices of nearest neighbours of each weight vector.
+  arma::Mat<size_t> weightNeighbourIndices(populationSize, neighbourhoodSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    // To temporarily store the distance between weights(i) and each other weights.
+    arma::vec distances(populationSize);
+    for (size_t j = 0; j < populationSize; j++)
+    {
+      distances(j) = 0;
+      for (size_t w = 0; w < numObjectives ; w++)
+      {
+        distances(j) += std::pow(weights[i](w) - weights[j](w), 2);
+      }
+      distances(j) = std::sqrt(distances(j));
+    }
+    arma::uvec sortedIndices = arma::stable_sort_index(distances, "descend");
+    for (size_t iter = 1; iter <= neighbourhoodSize; iter++)
+      weightNeighbourIndices(i, iter-1) = sortedIndices(iter);
+  }
+
+  // 1.3 Random generation of the initial population.
+  std::vector<MatType> population(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    population[i] = arma::randu<MatType>(iterate.n_rows, iterate.n_cols)
+        - 0.5 + iterate;
+
+  // 1.3 F-value initialisation for the population.
+  std::vector<arma::vec> FValue(populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+    FValue[i].resize(numObjectives);
+  EvaluateObjectives(population, objectives, FValue);
+
+  // 1.4 Initialize the ideal point z.
+  arma::vec idealPoint(numObjectives);
+  idealPoint.fill(-1/.0);
+
+  terminate |= Callback::BeginOptimization(*this, objectives, iterate, callbacks...);
+  // 2 The main loop.
+  for (size_t i = 0; i < populationSize && !terminate; i++)
+  {
+    terminate |= Callback::StepTaken(*this, objectives, iterate, callbacks...);
+
+    // To generate random numbers.
+    std::default_random_engine generator;
+
+    // Distribution to choose two random parents for mutation.
+    std::uniform_int_distribution<size_t> distribution(0, populationSize-1);
```

Comment:
What about using `arma::randi` that way it would be easier for a user to control the random number generator.

Reply:
Yes, I will replace it with `arma::randi` but what does the user having control over the random number generator exactly mean?

Reply:
Like I can use `arma_rng::set_seed(value)` to set the seed, in case I want to start with a specifc seed.

Reply:
When I replaced `distribution` with `arma::randi`, SchafferN1Test failed, what should I do? To make sure that the implementation has no problem and the tests passing is not a fluke, I tested it with another test function, [Kursawe Function](https://en.wikipedia.org/wiki/File:Kursawe_function.pdf). It gave points with values around -19.45 and 0.44, which seems about correct (but points around this one are not plotted in the graph I saw on Wikipedia, so I can't really say). I also checked  for [Binh and Korn Function](https://en.wikipedia.org/wiki/File:Binh_and_Korn_function.pdf) which gave correct answer around 0.04 and 48. I feel that either there is some problem in the implementation or MOEA/D does not work well when there is a discontinuous front, I am looking for any errors in the implementation.

Reply:
Maybe we slightly increase the population size to make the test more stable, if that makes sense?

Reply:
Actually I had initially kept population size to be 2000 and on @say4n's suggestion, I decided to look at the paper and found the recommended value to be around 150. That was the reason I changed it. I have found [another MOEA/D implementation](https://github.com/esa/pagmo/blob/master/src/algorithm/moea_d.cpp) and I am comparing it to mine. Will keep you updated.

Reply:
I guess you test against the same problem/function?

Reply:
Can you please elaborate? If you mean I should test for the same problems the other implementation has tested for, I am trying to comprehend their testing methodology. It is not very straight-forward.

Reply:
I have modified the SchafferN1Test a little bit. If a negative value is encountered, it checks if it's absolute value is less than the absolute value of the minimum positive value encountered, if it is, it does not report an error. Is this allowed? I tried finding errors in the implementation but could not, so this was the only way out.

Reply:
Hi @zoq, can you review this please? I am very excited to get this PR merged. Thanks!

Reply:
Absolutely, I'll set this on my list for later today.